platform/linux-generic/arch/x86/odp_sysinfo_parse.c

```diff
@@ -15,34 +15,54 @@ int cpuinfo_parser(FILE *file, system_info_t *sysinfo)
 	char str[1024];
 	char *pos, *pos_end;
 	double ghz = 0.0;
+	double mhz = 0.0;
 	uint64_t hz;
 	int id = 0;
+	bool freq_set = false;
 
 	strcpy(sysinfo->cpu_arch_str, "x86");
 	while (fgets(str, sizeof(str), file) != NULL && id < CONFIG_NUM_CPU) {
 		pos = strstr(str, "model name");
 		if (pos) {
+			freq_set = false;
+
 			/* Copy model name between : and @ characters */
 			pos     = strchr(str, ':');
 			pos_end = strchr(str, '@');
-			if (pos == NULL || pos_end == NULL)
+			if (pos == NULL)
 				continue;
 
-			*(pos_end - 1) = '\0';
+			if (pos_end != NULL)
+				*(pos_end - 1) = '\0';
+
 			strncpy(sysinfo->model_str[id], pos + 2,
 				MODEL_STR_SIZE - 1);
 
 			if (sysinfo->cpu_hz_max[id]) {
+				freq_set = true;
 				id++;
 				continue;
 			}
 
 			/* max frequency needs to be set */
-			if (sscanf(pos_end, "@ %lfGHz", &ghz) == 1) {
+			if (pos_end != NULL &&
+			    sscanf(pos_end, "@ %lfGHz", &ghz) == 1) {
 				hz = (uint64_t)(ghz * 1000000000.0);
-				sysinfo->cpu_hz_max[id] = hz;
+				sysinfo->cpu_hz_max[id++] = hz;
+				freq_set = true;
+			}
+		} else if (!freq_set &&
+			   strstr(str, "bogomips") != NULL) {
+			pos     = strchr(str, ':');
+			if (pos == NULL)
+				continue;
+
+			if (sscanf(pos + 2, "%lf", &mhz) == 1) {
+				/* On typical x86 BogoMIPS is freq * 2 */
```

Comment:
First 2 patches are ok. But I'm not sure about that statement. My i7-4710HQ CPU @ 2.50GHz has 4988.28 bogomips. Which is about double. But frequency jumps from 1Ghz to 2.5Ghz. If we say odp app rolls in busy loop than it looks like true. At least it's better to set up some dummy number. Other solution might be to calculate frequency with odp_cpu_cycles() (rdtsc) and clock_gettime().

Reply:
It's about max frequency, of course.